lms/services/jstor/service.py

```diff
@@ -22,7 +23,16 @@ class JSTORService:
     """Used when no DOI prefix can be found."""
 
     # pylint: disable=too-many-arguments
-    def __init__(self, api_url, secret, enabled, site_code, http_service: HTTPService):
+    def __init__(
+        self,
+        api_url,
+        secret,
+        enabled,
+        site_code,
+        http_service: HTTPService,
+        tracking_user_id: Optional[str],
+        tracking_user_agent: Optional[str],
```

Comment:
If these are optional, maybe they should be defaulted to `None`?

Reply:
The number of arguments here is getting a bit intense. I have a suggestion for cutting this down which is to use a similar approach to the `VitalSourceClient`. We could accept a dict of additional headers to pass to JSTOR without having each one specified as a new variable. This might look something like this:

```python
def service_factory(_context, request):
    ...
    return JSTORService(
        ...,
        headers={
            "Tracking-User-ID": request.lti_user.h_user.username
            if request.lti_user
            else None,
            "Tracking-User-Agent": request.headers.get("User-Agent", None),
        }
    )

...

class JSTORService:
    def __init__(self, api_url, secret, enabled, site_code, headers):
        ...
        self._http = HTTPService()
        self._http.session.headers = headers
```

By setting these headers in the session, they will be sent on every request. This would mean less code was required overall, and we keep he key and value of the headers next to each other rather than separated between two files.

This will require a change in the tests as we aren't passing the session any more.

Reply:
> If these are optional, maybe they should be defaulted to None?

Makes sense.

> The number of arguments here is getting a bit intense.

I see having many optional arguments as mainly a problem when they are all positional, leading to readability/refactoring hazards if some call site passes them positionally. Keyword-only args is one way to avoid that hazard, and similar to what one would do in other languages with an options hash/dict/struct at the end.

> We could accept a dict of additional headers to pass to JSTOR without having each one specified as a new variable

This is fine, as long as the logic remains strictly pass-through (eg. we don't decide to, say, limit tracking info to certain API calls) and the mechanism for passing them to the API doesn't change (eg. from headers to query params). I suspect that both these conditions will hold in this case.

> By setting these headers in the session, they will be sent on every request. 

Is there a downside to creating an extra `HTTPService` in the service? Is there anything from the shared service that would be lost in the process or some efficiency cost to be aware of?

Reply:
>Is there a downside to creating an extra HTTPService in the service? Is there anything from the shared service that would be lost in the process or some efficiency cost to be aware of?

No not really. The only efficiency it brings is if more than one service were to talk to the same site, then the SSL handshake would be saved. As we should be the only one talking to JSTOR, this is fine.

We can't use the global HTTP session, as putting headers in there would mess with everything that does use the global one.

It's probably a moot point anyway, as the "global" session is request scoped, so I doubt we get much from it anyway unless making more than one call during the same request.